### PR TITLE
C++: Fix some FPs in `cpp/missing-check-scanf` (second attempt)

### PR DESCRIPTION
**Pull request by MathiasVP**
#16009 gave good results, but was unfortunately closed due to performance problems in DCA. Initially, I had concluded that GVN was causing `ensuresEq(test, e, ..., block, ...)` to hold for too many (`e, block`) pairs and thus causing a slowdown.

After working through some performance problems in Ben's alternative PR https://github.com/github/codeql/pull/17907 I realized that the performance problem was actually coming from an existing problem in the `unary_simple_comparison_eq` predicate: currently, on `main`, there is a very large QLDoc and some very ad-hoc conjunctions in the second disjunct of `unary_simple_comparison_eq` to reduce the size of the predicate. However, it turns out that this is too doing a very good job.

Instead, [the second commit](db380692904c9ed545f7fc23383c2cda99efd46d) is restricting the predicate _properly_ by pruning the set down to only those tests that are used in unary comparisons that we actually care about.

Commits a40c1d50b8683a02b81f6a0dd1cd274c3fa778f7 and db380692904c9ed545f7fc23383c2cda99efd46d are taken from #16009 which @jketema has already looked at (although I don't blame you for wanting to re-review it if you want to. Let me know if you want me to split up the complex commit again like in #16009). The only new thing is the pruning in 442968c3c2b685b2b2217c3374dc764653d286f2

The commits that were tested at the latest DCA is consistent with the HEAD of this branch. I just force-pushed to clean up tests remove some spurious imports I had when experimenting with stuff.

### CODE REVIEW

**Comment by MathiasVP** [comment]
Bahh.... WAY too many join orders are broken by this and I don't feel like fixing them all 😪 I'll go back to keeping the non-GVN interface

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 1]
```diff
@@ -59,7 +60,70 @@ class MatchValue extends AbstractValue, TMatchValue {
...
+  override CompareGTInstruction cmp;
+}
+
+private class CompareLEValueNumber extends CompareValueNumber {
+  override CompareLEInstruction cmp;
+}
+
+private class CompareGEValueNumber extends CompareValueNumber {
+  override CompareGEInstruction cmp;
+}
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:106]: Are these all possible cases?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 2]
```diff
@@ -59,7 +60,70 @@ class MatchValue extends AbstractValue, TMatchValue {
...
+
+/**
+ * A value number such that at least one of the instructions is an
+ * insruction that is used in a `SwitchInstruction`'s expression.
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:110]:
```suggestion
 * A value number such that at least one of the instructions provides
 * the integer value controlling a  `SwitchInstruction`.
```

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 3]
```diff
@@ -59,7 +60,70 @@ class MatchValue extends AbstractValue, TMatchValue {
...
+ */
+private class ScrutineeValueNumber extends ValueNumber {
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:112]: Maybe `ControllingValueNumber`?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 4]
```diff
@@ -919,36 +999,56 @@ private class BuiltinExpectCallInstruction extends CallInstruction {
...
+}
+
+private class LogicalNotValueNumber extends ValueNumber {
+  LogicalNotInstruction instr;
+
+  LogicalNotValueNumber() { this.getAnInstruction() = instr }
+
+  ValueNumber getUnary() { result.getAnInstruction() = instr.getUnary() }
+}
+
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1028]: Maybe group these near the top with the `CompareValueNumber`?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 5]
```diff
@@ -827,82 +894,95 @@ private predicate unary_compares_eq(
   // ((test is `areEqual` => op == const + k2) and const == `k1`) =>
   // test is `areEqual` => op == k1 + k2
   inNonZeroCase = false and
-  exists(int k1, int k2, ConstantInstruction const |
+  exists(int k1, int k2, Instruction const |
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:897]: Why this change?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 6]
```diff
@@ -897,7 +965,7 @@ private predicate unary_simple_comparison_eq(
   not test.isGLValue() and
   not simple_comparison_eq(test, _, _, _, _) and
   not simple_comparison_lt(test, _, _, _) and
-  not test = any(SwitchInstruction switch).getExpression() and
+  op = test.getExpressionOperand() and
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll]: Although this is changed in a later commit. What is the reason behind this change?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 7]
```diff
@@ -1007,24 +1107,25 @@ private predicate compares_lt(
   exists(boolean isGe | isLt = isGe.booleanNot() | compares_ge(test, left, right, k, isGe, value))
   or
   /* (x is true => (left < right + k)) => (!x is false => (left < right + k)) */
-  exists(AbstractValue dual | value = dual.getDualValue() |
-    compares_lt(test.(LogicalNotInstruction).getUnary(), left, right, k, isLt, dual)
+  exists(AbstractValue dual |
+    value = dual.getDualValue() and
+    compares_lt(test.(LogicalNotValueNumber).getUnary(), left, right, k, isLt, dual)
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1112]: Why the switch from `exists( ... | ... | ...)` to `exists( ... | ... and ...)`?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 8]
```diff
@@ -1007,24 +1107,25 @@ private predicate compares_lt(
...
   or
   complex_lt(test, op, k, isLt, value)
   or
   /* (x is true => (op < k)) => (!x is false => (op < k)) */
-  exists(AbstractValue dual | value = dual.getDualValue() |
-    compares_lt(test.(LogicalNotInstruction).getUnary(), op, k, isLt, dual)
+  exists(AbstractValue dual |
+    value = dual.getDualValue() and
+    compares_lt(test.(LogicalNotValueNumber).getUnary(), op, k, isLt, dual)
   )
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1126]: Why the switch from `exists( ... | ... | ...)` to `exists( ... | ... and ...)`?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 9]
```diff
@@ -1007,24 +1107,25 @@ private predicate compares_lt(
...
   or
-  exists(int k1, int k2, ConstantInstruction const |
+  exists(int k1, int k2, Instruction const |
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1128]: Why this change?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 10]
```diff
@@ -926,29 +994,47 @@ private class BuiltinExpectCallInstruction extends CallInstruction {
...
     or
-    cmp instanceof CompareEQInstruction and
+    cmp instanceof CompareEQValueNumber and
     value.getDualValue() = innerValue
   )
 }
 
 private predicate complex_eq(
-  CompareInstruction cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
+  ValueNumber cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1051]: Why not `CompareValueNumber`?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 11]
```diff
@@ -1081,15 +1164,15 @@ private predicate unary_simple_comparison_lt(
 }
 
 private predicate complex_lt(
-  CompareInstruction cmp, Operand left, Operand right, int k, boolean isLt, AbstractValue value
+  ValueNumber cmp, Operand left, Operand right, int k, boolean isLt, AbstractValue value
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1181]: Why not `CompareValueNumber`?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 12]
```diff
@@ -1099,7 +1182,7 @@ private predicate complex_lt(
 // left - x < right + c => left < right + (c+x)
 // left < (right - x) + c => left < right + (c-x)
 private predicate sub_lt(
-  CompareInstruction cmp, Operand left, Operand right, int k, boolean isLt, AbstractValue value
+  ValueNumber cmp, Operand left, Operand right, int k, boolean isLt, AbstractValue value
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1199]: Why not `CompareValueNumber`?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 13]
```diff
@@ -1149,7 +1232,7 @@ private predicate sub_lt(Instruction test, Operand left, int k, boolean isLt, Ab
 // left + x < right + c => left < right + (c-x)
 // left < (right + x) + c => left < right + (c+x)
 private predicate add_lt(
-  CompareInstruction cmp, Operand left, Operand right, int k, boolean isLt, AbstractValue value
+  ValueNumber cmp, Operand left, Operand right, int k, boolean isLt, AbstractValue value
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1249]: Why not `CompareValueNumber`?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 14]
```diff
@@ -1217,7 +1300,7 @@ private predicate add_lt(Instruction test, Operand left, int k, boolean isLt, Ab
 // left - x == right + c => left == right + (c+x)
 // left == (right - x) + c => left == right + (c-x)
 private predicate sub_eq(
-  CompareInstruction cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
+  ValueNumber cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1317]: Why not `CompareValueNumber`?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 15]
```diff
@@ -1272,7 +1355,7 @@ private predicate unary_sub_eq(
 // left + x == right + c => left == right + (c-x)
 // left == (right + x) + c => left == right + (c+x)
 private predicate add_eq(
-  CompareInstruction cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
+  ValueNumber cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1372]: Why not `CompareValueNumber`?

**Review by jketema — COMMENTED** [review]
Comments on the first two commits. There's a few comments to occur in several places. Feel free to only react to one of those if you think nothing needs to be changed.

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 6 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll]: Oh, right. That was a bug I temporarily introducered and forgot to remove from the commit. The `not` shouldn't have been removed.

But, as you noted, this code is removed in the next commit so 🤷‍♂️

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:106]: Yes, these are all the subtypes of `RelationalIntruction`

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 5 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:897]: Oh, that wasn't actually needed here, no. I've got an another upcoming PR to add value numbering to the `int_value` predicate (which currently takes a `ConstantInstruction`) in order to detect more constant cases, and that would need this particular change as part of it.

RIght now, this type relaxation doesn't change anything on this PR as the type of `const` is restricted by the call to `int_value`. I'm happy to undo this if you prefer, though

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:897]: It's fine to leave it as it is.

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 7 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1112]: Oh, that's a mistake. I think I was quick-evaling `... and ...` at some point during development since it's still not possible to quick eval `... | ...` 😭

I don't have a strong preference on `... | ...` vs. `... and ...` in this case since my normal motivation for preferring `... | ...` over `... and ...` is to save parentheses 😄

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 8 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1126]: Same as above

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 9 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1128]: Same as in https://github.com/github/codeql/pull/17938#discussion_r1843777836

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 7 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1112]: Let's just change this back to `exists( ... | ... | ...)` then.

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 8 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll]: Fixed in c80bea33e527fe2a149c2cf54f2a23138eef370c

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 7 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll]: Fixed in c80bea33e527fe2a149c2cf54f2a23138eef370c

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 3 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:112]: I like that (mostly because I hate the word "scrutinee" 😭). The only thing I dislike about `Controlling` is that it doesn't make it clear that it's specifically for switches (and I think "scrutinee" is a slightly more standard way of describing the expression in a switch.

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:112]: So I was just going by how the comments in the IR refer to them. The standard just seems to call it a "condition" so how about `SwitchCondition`?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll]: Fixed in 5f9d3d7b09270c33eed40a0994344696d8c92ad2

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 16]
```diff
@@ -827,82 +912,95 @@ private predicate unary_compares_eq(
...
+  )
+  or
+  // If `y` is a relevant unary comparison and `y = x` then so is `x`.
+  not op.isDefinitionInexact() and
+  exists(CopyInstruction copy |
+    isRelevantUnaryComparisonOperand(unique( | | copy.getAUse())) and
+    op = copy.getSourceValueOperand()
+  )
+  or
+  // If phi(x1, x2) is a relevant unary comparison then so is `x1` and `x2`.
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:975]:
```suggestion
  // If phi(x1, x2) is a relevant unary comparison then so are `x1` and `x2`.
```

**Review by jketema — COMMENTED** [review]
This feels a lot less magical than what was there before. One small comment.

**Comment by jketema** [comment]
DCA doesn't show any alert differences. How about MRVA?

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll (hunk 10 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1051]: We end up calling into the general `compares_eq` predicate, and that has cases where it's not a `CompareValueNumber`. For example in the case where it's a `LogicalNotValueNumber`. I'm actually surprised why this wasn't broken on `main` already, but I haven't been able to produce a counter-example 🤔 In any case, I understand why _this_ works so it seems slightly more safe

**Comment by MathiasVP** [comment]
Something odd going on when testing this locally. The DB/commit pair that seemed to work on DCA is now exploding locally 😰 I'm debugging this now!

**Comment by MathiasVP** [comment]
I've managed to figure out why DCA runs fine, but local runs hit a tuple explosion: The problem is that the `IRGuardCondition::ensuresEq/4` is never actually evaluated in the `security-and-quality` suite. Instead queries use, `GuardCondition::ensuresEq/4` (i.e., the AST interface version which for some reason re-implements the functionality so that it doesn't depend on the `IRGuardCondition::ensuresEq/4` predicate) which doesn't have the same problem (because it's restricted to instruction for which `getUnconvertedResultExpression` has a result).

But because they're all in the same cached stage they get evaluated together when you don't specify `--expect-discarded-cache` (because the other predicate may be needed in some subsequent run that the evaluator isn't aware of). But when the flag is present the evaluator knows that the cache will be discarded and thus chooses not to evaluate it (at least, that's my mental model of `--expect-discarded-cache`).

Obviously we may want to call `IRGuardCondition::ensuresEq/4` in the future, so the size of this predicate should be reduced so that we can evaluate it. I've got a [fix](https://github.com/github/codeql/commit/61fb0cbee63b5459e3440b553a83c7a0ecbacd9f) that I'm testing now. Locally, this makes the `IRGuardCondition::ensuresEq/4` feasible so I'm hopeful 🤞 But I'm running DCA on it now!

**Comment by jketema** [comment]
Did you observe any alert differences in MRVA?

**Comment by MathiasVP** [comment]
I can run a MRVA diff query on a couple of queries that use `GuardCondition`s, sure 👍

We've seen good results on the Microsoft side already, though.

**Comment by MathiasVP** [comment]
> I'd mostly be interested to know if we observe changes to the `cpp/missing-check-scanf` results in MRVA.

MRVA results for `cpp/missing-check-scanf` is in: https://gist.github.com/MathiasVP/2dd7eaf2838a5a2226db078a0e54ce6e

TLDR: Two FPs removed on top 1000 🎉

**Comment by jketema** [comment]
Thanks. I think this is ok. I wonder if we do need a change note, because of the changed behavior related to globals and static locals?

**Comment by MathiasVP** [comment]
Good idea! I'll add that 👍

File: cpp/ql/lib/change-notes/2024-11-18-guard-conditions.md [hunk 17]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* The `Guards` library ("semmle.code.cpp.controlflow.Guards") has been improved to recognize more guard conditions. Additionally, the guards library no longer considers guards in static local initializers or global initializers as `GuardCondition`s.
```
Comment (jketema) [inline cpp/ql/lib/change-notes/2024-11-18-guard-conditions.md:4]:
```suggestion
* The `Guards` library (`semmle.code.cpp.controlflow.Guards`) has been improved to recognize more guard conditions. Additionally, the guards library no longer considers guards in static local initializers or global initializers as `GuardCondition`s.
```
